### DISCUSSION TITLE
Allow shell selection for entire CMake Tools infrastructure

### DISCUSSION THREAD

**Discussion by bjm9991** [microsoft/vscode-cmake-tools#3526] [Ideas]
There are many open MSYS2/MinGW issues in `vscode-cmake-tools` which all stem from the same problem:
_There is no way to select a shell in which CMake Tools runs_.

I (and many other) developers prefer to use MSYS2+MinGW or similar over MSVC and native toolchains because it provides a familiar experience for people coming from the Linux or embedded development space where we can rely on the existence of certain builtins, optimisations and features. There's also WSL of course, but I think the approach there is to treat it as a remote development environment (through the *cough* proprietary *cough* extension) - besides it doesn't build native Windows executables.

The kit specification feature is quite underpowered at the moment. One can specify compiler locations for example, but there is _so_ much more to a complete development environment than just the frontend language compilers. There's binutils, linkers, library search paths etc. Many of the problems in open issues around Kits stems from this fact - people are trying to point CMake to their "C compiler of choice" while the rest of their environment is completely out of whack. The kit structure _encourages_ people to fall into this trap and it is a bad design decision IMO. Operating systems typically come in "distributions" for a reason. One easy solution to this for the MSYS2 case would be to just allow us to specify which shell the tools run CMake in. You already let us specify a custom "Terminal" which can interact with MSYS2 bash no problem.

Through the `MSYSTEM` environment variable, you can spin up a MSYS2 shell geared for e.g. MinGW64 development, with the PATH correctly configured to include the prefix at C:\msys64\mingw64\bin and so on.

The current development workflow for someone using MSYS2 currently consists of either: running a terminal emulator and calling cmake manually (with IntelliSense kind of getting the idea most of the time); or specifying a complex and fragile array of variables in cmake-kits.json, CMakePresets.json and settings.json to get the whole thing to work.

**Comment by gcampbell-msft** [comment]
In an effort to minimize how many places we have to track these issues, I'm going to close this in favor of the various issues you mention.

In response to your comments about Kits, we would suggest that you favor using CMakePresets as this is the officially supported path by Kitware (developers of CMake) and we are prioritizing CMakePresets in this extension. 

Thanks!

**Comment by debugee** [comment]
is this done?

**Comment by bjm9991** [comment]
@gcampbell-msft Can you please explain to me how CMakePresets has anything to do with shell selection? It is frustrating when maintainers do not properly read these issues. Please reopen this.

**Reply by debugee** [reply]
I agree with you, please reopen this.
cmake-tools run cmake do not use vscode shell config.
vscode debug\task\shell can select by settings.json file.
cmake-tool just set cmake-tools-kits.json 's env setting.
though we can set cmake-tools-kits.json file env when use msys2,and work will
but cmake said:(https://cmake.org/cmake/help/latest/variable/CMAKE_SYSTEM_NAME.html)

Platform-specific notes:

MSYS2's msys/cmake package (/usr/bin/cmake) works only under MSYSTEM=MSYS environments, with system name MSYS. Under other environments like MSYSTEM=MINGW64, use another package such as mingw64/mingw-w64-x86_64-cmake (/mingw64/bin/cmake), which targets MSYSTEM=MINGW64 with system name Windows.
Cygwin's cmake package (/usr/bin/cmake) uses system name CYGWIN. A non-cygwin CMake on Windows (e.g. $PROGRAMFILES/CMake/bin/cmake) uses system name Windows even when it runs under a Cygwin environment.

**Comment by gcampbell-msft** [comment]
@benjamin-branchware My response was in 2 parts, first, I closed the discussion in favor of tracking the various issues that you referenced are already open in the repository. However, I'm happy to keep open an overall issue like this, if you could convert this or create a new issue for it. We prioritize using the issues list to prioritize work rather than the discussions pane.

The second part of my response, the presets content, was regarding your large paragraph referencing things that are lacking in the kits feature. This is not prioritized as we are focused primarily on Presets. 

I agree that it sounds like the shell selection could be very beneficial for folks, so I'd be happy to keep an issue open provided you convert this or create a new one.

**Reply by debugee** [reply]
you said much，you said nothing

**Reply by gcampbell-msft** [reply]
We're happy to track and investigate this request if you create an issue for us to track. 

I've created an issue from this discussion here: #4362
